### PR TITLE
Add REST API GET, POST, PATCH tests for cloud storage

### LINKED ISSUE #3996: Add support for S3 compatible storage providers (eg MinIO)
### My actions before raising this issue
- [x] Read/searched [the docs](https://github.com/opencv/cvat/tree/master#documentation)
- [x] Searched [past issues](/issues)

### Expected Behaviour

The new clowd storage feature is great but we have our Data on a private s3 compatible storage. There are multiple offerings for such storage on the market and it would be great if those would be usable as well. 

### Current Behaviour

Currently the AWS S3 cloud storage provider only supports AWS hosted S3 buckets.

### Possible Solution

In essence we need a way to configure endpoint_url for the boto3 resource when adding a cloud storage.

### PR DESCRIPTION
**Pull request by Marishka17**
### Motivation and context

PR contains REST API tests for cloud storage functionality
Resolves #3996
### Checklist

- [x] I submit my changes into the `develop` branch
- [ ] I have added a description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
~~- [ ] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly~~
- [x] I have added tests to cover my changes
~~- [ ] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))~~
~~- [ ] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))~~

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2022 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

**Comment by nmanovic** [comment]
@Marishka17 , could you please merge develop and share your experience about the process? I see that you have conflicts with cvat_data.tar.bz2, cvat_db.sql.

**Comment by nmanovic** [comment]
@kirill-sizov , could you please review the patch and provide your suggestions.

**Comment by Marishka17** [comment]
> @Marishka17 When I run these tests locally they fails, for example this test failed due to status code is `400`
> 
> ```
> FAILED test_cloud_storages.py::TestPostCloudStorage::test_sandbox_user_create_cloud_storage[user-True]
> ```
> 
> Am I doing something wrong? Maybe I need to change something in my setup?

You need to export `TESTING` environment variable and run docker containers with `-f docker-compose.testing.yml` option.

File: tests/rest_api/test_cloud_storages.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        if is_allow:
+            self._test_can_see(username, storage_id, cloud_storage, org=org)
+        else:
+            self._test_cannot_see(username, storage_id, org=org)
+
+
+    @pytest.mark.parametrize('org_id', [2])
+    @pytest.mark.parametrize('storage_id', [2])
+    @pytest.mark.parametrize('role, is_owner, is_allow', [
+        #('worker',       True,   True),
```
Comment (kirill-sizov) [inline tests/rest_api/test_cloud_storages.py:47]: Is there any reason to skip this test?

File: cvat/apps/engine/serializers.py [hunk 2]
```diff
@@ -1080,8 +1086,9 @@ def update(self, instance, validated_data):
             'specific_attributes': parse_specific_attributes(instance.specific_attributes)
         }
         storage = get_cloud_storage_instance(cloud_provider=instance.provider_type, **details)
+        is_testing = os.environ.get('TESTING', False)
```
Comment (nmanovic) [inline cvat/apps/engine/serializers.py:1089]: @Marishka17 , I don't think that we need to have the check in production code. Need to find another idea.

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -74,12 +74,13 @@ jobs:
       - name: Running REST API tests
         env:
           API_ABOUT_PAGE: "localhost:8080/api/server/about"
+          TESTING: "True"
```
Comment (nmanovic) [inline .github/workflows/main.yml:77]: Need to revert all these changes. We are testing a real instance on CVAT.

**Comment by Marishka17** [comment]
@kirill-sizov, Could you please help and resolve conflicts with assets?

**Comment by Marishka17** [comment]
@nmanovic, I've changed approach with test cloud storage functionality. Main idea is use [Multi-Cloud Object
Storage](https://min.io/). Now, before running REST API tests MinIO server is started and MinIO client creates several buckets locally that are used in tests. 
I'm removing WIP status now because I've mostly done all (maybe I will do only some cosmetics edits) and want to be sure that tests will be passed on CI as locally after resolving conflicts.

**Comment by kirill-sizov** [comment]
> @kirill-sizov, Could you please help and resolve conflicts with assets?

Yes sure, I will resolve conflicts.

File: tests/rest_api/setup [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+# create MINIO buckets with data
+MC_DIR="${HOME}/Downloads"
+MC_PATH="${MC_DIR}/mc"
+if [ ! -e ${MC_PATH} ]
+then
+    echo "Downloading mc tool"
+    wget -P ${MC_DIR} https://dl.min.io/client/mc/release/linux-amd64/mc
+    chmod +x ${MC_PATH}
+    alias mc=${MC_PATH}
```
Comment (kirill-sizov) [inline tests/rest_api/setup:26]: Such an alias could potentially conflict with popular tool Midnight Commander. I would suggest just use other way for using Minio Client, example:
```
${MC_PATH} alias set ${MINIO_ALIAS} ${MINIO_HOST} ${MINIO_ACCESS_KEY} ${MINIO_SECRET_KEY}
```

Comment (kirill-sizov) [inline tests/rest_api/setup:26]: Such an alias could potentially conflict with popular tool Midnight Commander. I would suggest just use other way for using Minio Client, example:
```
${MC_PATH} alias set ${MINIO_ALIAS} ${MINIO_HOST} ${MINIO_ACCESS_KEY} ${MINIO_SECRET_KEY}
```

File: tests/rest_api/setup [hunk 5]
```diff
@@ -0,0 +1,43 @@
+#!/bin/bash
```
Comment (kirill-sizov) [inline tests/rest_api/setup:1]: This setup file can only be applied on Linux. This makes it impossible to develop tests for cloud storage from Windows/MacOS machine. I see few ways for resolving this problem:

1) Fast: mentioning about such inconvenience in README

2) May take a long time: find out a solution that can be applied on Linux, MacOS and on Windows as well.

3) Compromise between previous two: provide instructions for creating local MINIO buckets on different operating systems.

Comment (kirill-sizov) [inline tests/rest_api/setup:1]: This setup file can only be applied on Linux. This makes it impossible to develop tests for cloud storage from Windows/MacOS machine. I see few ways for resolving this problem:

1) Fast: mentioning about such inconvenience in README

2) May take a long time: find out a solution that can be applied on Linux, MacOS and on Windows as well.

3) Compromise between previous two: provide instructions for creating local MINIO buckets on different operating systems.

File: tests/rest_api/test_cloud_storages.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+        if is_allow:
+            self._test_can_create(username, self._SPEC, org=org)
+        else:
+            self._test_cannot_create(username, self._SPEC, org=org)
+
+    @pytest.mark.parametrize('org_id', [2])
+    @pytest.mark.parametrize('role, is_allow', [
+        ('maintainer', True),
+        ('supervisor', False),
```
Comment (kirill-sizov) [inline tests/rest_api/test_cloud_storages.py:113]:
```suggestion
        ('owner', True), ('maintainer', True),
        ('worker', False), ('supervisor', False),
```

Comment (kirill-sizov) [inline tests/rest_api/test_cloud_storages.py:113]:
```suggestion
        ('owner', True), ('maintainer', True),
        ('worker', False), ('supervisor', False),
```

File: tests/rest_api/test_cloud_storages.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+            exclude_paths=self._EXCLUDE_PATHS) == {}
+
+    def _test_cannot_create(self, user, spec, **kwargs):
+        response = post_method(user, 'cloudstorages', spec, **kwargs)
+
+        assert response.status_code == HTTPStatus.FORBIDDEN
+
+
+    @pytest.mark.parametrize('group, is_allow', [
+        ('user', True),
```
Comment (kirill-sizov) [inline tests/rest_api/test_cloud_storages.py:99]:
```suggestion
        ('user', True), ('worker', False)
```

Comment (kirill-sizov) [inline tests/rest_api/test_cloud_storages.py:99]:
```suggestion
        ('user', True), ('worker', False)
```

**Review by kirill-sizov — APPROVED** [review]
Looks good to me.

**Review by nmanovic — APPROVED** [review]
@Marishka17 , great job!

**Pull request merged**